File: backend/youjp/ws/protocol.py
"""Mensajes de texto del WebSocket.

Los esquemas viven aqui y de aqui salen los tipos de TypeScript de la extension
(``scripts/gen_ts_types.py``), para que el protocolo no pueda desincronizarse
entre los dos lados sin que alguien se entere.

Las tramas de audio no pasan por aqui: van en binario, ver :mod:`youjp.ws.codec`.
"""

from __future__ import annotations

from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

# --------------------------------------------------------------------------
# cliente -> servidor
# --------------------------------------------------------------------------


class SessionStart(BaseModel):
    type: Literal["session.start"] = "session.start"
    video_id: str = ""
    url: str = ""
    is_live: bool = False
    media_time_ms: int = 0
    source: str = "ja"
    target: str = "es"
    profile: str = "n4"


class SessionStop(BaseModel):
    type: Literal["session.stop"] = "session.stop"


class ControlFlush(BaseModel):
    """Seek, pausa o cambio de velocidad: se descarta el buffer y se reancla."""

    type: Literal["control.flush"] = "control.flush"
    reason: Literal["seek", "pause", "rate", "manual"] = "manual"
    media_time_ms: int = 0


class Ping(BaseModel):
    type: Literal["ping"] = "ping"
    t: int = 0


ClientMessage = Annotated[
    Union[SessionStart, SessionStop, ControlFlush, Ping],
    Field(discriminator="type"),
]


# --------------------------------------------------------------------------
# servidor -> cliente
# --------------------------------------------------------------------------


class SessionReady(BaseModel):
    type: Literal["session.ready"] = "session.ready"
    session_id: str
    asr_model: str
    device: str
    compute_type: str
    sample_rate: int
    frame_ms: int
    protocol_version: int = 1


class AsrPartial(BaseModel):
    """Texto en curso.

    ``committed`` ya no cambiara nunca: es lo que LocalAgreement ha confirmado y
    se puede pintar en firme. ``tentative`` se reemplaza entera en cada
    actualizacion, asi que se pinta en gris y no se le montan interacciones
    encima.
    """

    type: Literal["asr.partial"] = "asr.partial"
    committed: str
    tentative: str
    media_start_ms: int


class AsrFinal(BaseModel):
    type: Literal["asr.final"] = "asr.final"
    segment_id: int
    text: str
    media_start_ms: int
    media_end_ms: int
    reason: str
    latency_ms: float


class MtFinal(BaseModel):
    """Traducción de una frase ya cerrada.

    Llega por separado y después del ``asr.final`` correspondiente, y se enlaza
    con él por ``segment_id``. Así el japonés aparece en cuanto está listo, sin
    esperar al traductor: son dos trabajos en paralelo, no una cadena.
    """

    type: Literal["mt.final"] = "mt.final"
    segment_id: int
    text_es: str
    provider: str
    media_start_ms: int
    media_end_ms: int
    latency_ms: float


class MetricsTick(BaseModel):
    type: Literal["metrics.tick"] = "metrics.tick"
    end_to_end_ms_p50: float = 0.0
    end_to_end_ms_p95: float = 0.0
    whisper_processing_ms_p50: float = 0.0
    audio_buffer_ms: float = 0.0
    dropped_audio_chunks: int = 0
    partial_transcripts: int = 0
    final_transcripts: int = 0
    passes: int = 0
    skipped_silent: int = 0
    rejected: dict[str, int] = Field(default_factory=dict)
    translation_latency_ms_p50: float = 0.0
    translations: int = 0
    dropped_translations: int = 0
    mt_provider: str = "none"
    gpu_used_mb: float = 0.0
    gpu_total_mb: float = 0.0
    cpu_pct: float = 0.0
    ram_mb: float = 0.0


class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str
    fatal: bool = False


class Pong(BaseModel):
    type: Literal["pong"] = "pong"
    t: int = 0


ServerMessage = Union[
    SessionReady, AsrPartial, AsrFinal, MtFinal, MetricsTick, ErrorMessage, Pong
]


class ClientEnvelope(BaseModel):
    """Envoltorio solo para validar: pydantic necesita un modelo raiz para
    despachar por el campo ``type``."""

    message: ClientMessage


def parse_client_message(raw: str) -> ClientMessage:
    """Valida un mensaje de texto entrante.

    Lanza ``pydantic.ValidationError`` si no encaja con ningun tipo conocido.
    """
    return ClientEnvelope.model_validate_json(f'{{"message": {raw}}}').message

File: extension/src/protocol.ts
/**
 * Protocolo del WebSocket. Espejo de `backend/youjp/ws/protocol.py` y
 * `codec.py`.
 *
 * Mientras no haya generación automática de tipos, cualquier cambio aquí hay
 * que hacerlo también allí. El test `test_ws_codec.py` del backend cubre el
 * lado binario, que es donde un desajuste no da un error claro sino audio que
 * suena a ruido blanco.
 */

export const PROTOCOL_VERSION = 1;
export const SAMPLE_RATE = 16_000;
export const FRAME_MS = 100;
export const FRAME_SAMPLES = (SAMPLE_RATE * FRAME_MS) / 1000; // 1600
export const HEADER_SIZE = 16;

const MAGIC = 0xa5;
export const FLAG_LIVE = 1 << 0;
export const FLAG_DISCONTINUITY = 1 << 1;

// ---------------------------------------------------------------------------
// trama binaria
// ---------------------------------------------------------------------------

export interface FrameHeader {
  seq: number;
  mediaTimeMs: number;
  captureMs: number;
  flags: number;
}

/**
 * Antepone la cabecera de 16 B al PCM ya convertido a Int16 por el worklet.
 *
 * `setUint32(..., true)` en todos los campos: little endian. Sin ese `true`,
 * JavaScript escribe big endian por defecto y Python lee números absurdos —
 * el error más fácil de cometer en esta frontera.
 */
export function buildFrame(pcm: ArrayBuffer, header: FrameHeader): ArrayBuffer {
  const out = new ArrayBuffer(HEADER_SIZE + pcm.byteLength);
  const view = new DataView(out);

  view.setUint8(0, MAGIC);
  view.setUint8(1, PROTOCOL_VERSION);
  view.setUint8(2, header.flags);
  view.setUint8(3, 0); // reservado
  view.setUint32(4, header.seq >>> 0, true);
  view.setUint32(8, Math.max(0, Math.round(header.mediaTimeMs)) >>> 0, true);
  view.setUint32(12, Math.max(0, Math.round(header.captureMs)) >>> 0, true);

  new Uint8Array(out, HEADER_SIZE).set(new Uint8Array(pcm));
  return out;
}

// ---------------------------------------------------------------------------
// mensajes de texto
// ---------------------------------------------------------------------------

export interface SessionStart {
  type: 'session.start';
  video_id: string;
  url: string;
  is_live: boolean;
  media_time_ms: number;
  source: string;
  target: string;
  profile: string;
}

export interface SessionStop {
  type: 'session.stop';
}

export interface ControlFlush {
  type: 'control.flush';
  reason: 'seek' | 'pause' | 'rate' | 'manual';
  media_time_ms: number;
}

export type ClientMessage = SessionStart | SessionStop | ControlFlush | { type: 'ping'; t: number };

export interface SessionReady {
  type: 'session.ready';
  session_id: string;
  asr_model: string;
  device: string;
  compute_type: string;
  sample_rate: number;
  frame_ms: number;
  protocol_version: number;
}

/**
 * `committed` ya no cambiará nunca: es lo que LocalAgreement ha confirmado y se
 * puede pintar en firme. `tentative` se reemplaza entera en cada actualización,
 * así que va en gris y sin interacciones encima.
 */
export interface AsrPartial {
  type: 'asr.partial';
  committed: string;
  tentative: string;
  media_start_ms: number;
}

export interface AsrFinal {
  type: 'asr.final';
  segment_id: number;
  text: string;
  media_start_ms: number;
  media_end_ms: number;
  reason: string;
  latency_ms: number;
}

/**
 * Traducción de una frase ya cerrada.
 *
 * Llega por separado y después del `asr.final` correspondiente, enlazada por
 * `segment_id`. Así el japonés aparece en cuanto está listo sin esperar al
 * traductor: son dos trabajos en paralelo, no una cadena.
 */
export interface MtFinal {
  type: 'mt.final';
  segment_id: number;
  text_es: string;
  provider: string;
  media_start_ms: number;
  media_end_ms: number;
  latency_ms: number;
}

export interface Sense {
  /** Etiquetas de JMdict: `n`, `vs`, `adj-i`... */
  pos: string[];
  glosses_es: string[];
  glosses_en: string[];
}

export interface DictEntry {
  id: number;
  headword: string;
  readings: string[];
  common: boolean;
  freq_rank: number | null;
  senses: Sense[];
}

/**
 * Una unidad clicable del subtítulo.
 *
 * La entrada de diccionario viene incrustada en vez de un identificador para
 * pedirla después: una frase son unos pocos kilobytes por localhost, y a cambio
 * la tarjeta aparece en el mismo fotograma del clic.
 */
export interface Token {
  i: number;
  /** Índices sobre el texto de `asr.final`. */
  span: [number, number];
  surface: string;
  lemma: string;
  kana: string;
  romaji: string;
  pos: string[];
  pos_label: string;
  /** Cadena de conjugación: `["causativo", "formal", "pasado"]`. */
  chain: string[];
  clickable: boolean;
  entry: DictEntry | null;
}

export interface NlpTokens {
  type: 'nlp.tokens';
  segment_id: number;
  tokens: Token[];
  analysis_ms: number;
}

export interface MetricsTick {
  type: 'metrics.tick';
  end_to_end_ms_p50: number;
  end_to_end_ms_p95: number;
  whisper_processing_ms_p50: number;
  audio_buffer_ms: number;
  dropped_audio_chunks: number;
  partial_transcripts: number;
  final_transcripts: number;
  passes: number;
  skipped_silent: number;
  rejected: Record<string, number>;
  nlp_ms_p50: number;
  translation_latency_ms_p50: number;
  translations: number;
  dropped_translations: number;
  mt_provider: string;
  gpu_used_mb: number;
  gpu_total_mb: number;
  cpu_pct: number;
  ram_mb: number;
}

export interface ServerError {
  type: 'error';
  code: string;
  message: string;
  fatal: boolean;
}

export type ServerMessage =
  | SessionReady
  | AsrPartial
  | AsrFinal
  | MtFinal
  | NlpTokens
  | MetricsTick
  | ServerError
  | { type: 'pong'; t: number };
